### PR TITLE
added Slider and Popup Components to the library

### PR DESCRIPTION
**Pull request by sahandfarahani**
photo_2020-04-27 21 26 23
photo_2020-04-27 21 26 28
photo_2020-04-27 21 26 32
photo_2020-04-27 21 26 35
photo_2020-04-27 21 26 40
photo_2020-04-27 21 26 44

### CODE REVIEW

File: src/Containers/Popup.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+import styled from 'styled-components';
+import { LabelLayoutProps } from '@Layouts';
+
+export interface PopupProps extends LabelLayoutProps {
+    left?:string|number;
```
Comment (ralph-dev) [inline src/Containers/Popup.tsx:6]: Put the spaces after the ":"

File: src/Inputs/Slider.tsx [hunk 2]
```diff
@@ -1,80 +1,351 @@
...
+            const bar = barRef.current as HTMLElement;
+
+            const selectedBar = bar.firstChild as HTMLElement;
+            const finishThumb = bar.lastChild as HTMLElement;
+            const startThumb = finishThumb.previousElementSibling as HTMLElement;
+            const childrenbar = bar.nextElementSibling as HTMLElement;
+
+            const clientwidth = bar.clientWidth;
+
+            const maxinpixels = translateToPixels(max,clientwidth,max,min as number);
```
Comment (ralph-dev) [inline src/Inputs/Slider.tsx:77]: maxInPixels follow conventions

File: src/Inputs/Slider.tsx [hunk 3]
```diff
@@ -1,80 +1,351 @@
...
+            if (popup){
+                let popUpContainer = (thumb.firstChild as HTMLElement).firstChild as HTMLElement;
+                (popUpContainer.textContent as string|number) = translateToValue(newLeft, barwidth, max, min as number, step as number);    
+            }
+            
+            thumb.style.left = calculatePosition(newLeft, barwidth, max, min as number, step as number);
+
+          } 
+        function onMouseUp() {
+            document.removeEventListener('mouseup', onMouseUp);
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:171]: Why is there an event listener on the body and not the component itself?

File: src/Containers/Popup.tsx [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+export interface PopupProps extends LabelLayoutProps {
+    left?:string|number;
+    top?:string|number;
+    width?:string|number;
+    height?:string|number;
+    popup?:boolean;
+}
+
+export const Popup : React.FC<PopupProps> = ({
```
Comment (fpunny) [inline src/Containers/Popup.tsx:13]: This shouldn't be shared since it's only used by one component

File: src/Inputs/Slider.tsx [hunk 5]
```diff
@@ -1,80 +1,351 @@
...
-        height: 36px;
-        width: 16px;
-        border-radius: 50%;
-        background: #ffffff;
-        cursor: pointer;
-    }
-`;
-
-export default Slider;
+import React,{useLayoutEffect,useRef, RefObject,MouseEvent} from 'react';
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:1]: For types, use React.xxxx

File: src/Inputs/Slider.tsx [hunk 6]
```diff
@@ -1,80 +1,351 @@
...
+    min,
+    max,
+    popupleft=-17,
+    popuptop=-51,
+    popupwidth="auto",
+    popupheight=30,
+    ...props
+}): React.ReactElement =>{  
+
+    const barRef = useRef() as RefObject<HTMLDivElement>;
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:50]: useRef<HTMLDivElement>(null)

File: src/Inputs/Slider.tsx [hunk 7]
```diff
@@ -1,80 +1,351 @@
...
+    
+    const translateToPixels = (theValue:number|undefined, scale:number|undefined, maximum:number|undefined, minimum:number) : string =>{
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:52]: This is not consistent code style

File: src/Inputs/Slider.tsx [hunk 8]
```diff
@@ -1,80 +1,351 @@
...
+        const maxmindifference = (maximum as number - minimum as number) as number;
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:53]: Wrap everything with an if check rather than forcing types

File: src/Inputs/Slider.tsx [hunk 9]
```diff
@@ -1,80 +1,351 @@
...
+        const pixelTranslator = scale as number / maxmindifference;
+
+        return ((theValue as number - minimum as number) * pixelTranslator) + "px";
+    }
+
+    const translateToValue=(theValue:number, scale:number, maximum:number|undefined, minimum:number, step:number)=>{
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:59]: Some of these calculations can be memo-ized

File: src/Inputs/Slider.tsx [hunk 10]
```diff
@@ -1,80 +1,351 @@
...
+    }
+
+    const calculatePosition = (theValue:number, scale:number, maximum:number|undefined, minimum:number, step:number) =>{
+        return translateToPixels(translateToValue(theValue,scale,maximum,minimum,step),scale,maximum as number|undefined,minimum);
+    }
+    useLayoutEffect(():void => {
+        if (null !== barRef.current){
+            const bar = barRef.current as HTMLElement;
+
+            const selectedBar = bar.firstChild as HTMLElement;
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:70]: This is bad, use refs

File: src/Inputs/Slider.tsx [hunk 11]
```diff
@@ -1,80 +1,351 @@
...
+            const startThumb = finishThumb.previousElementSibling as HTMLElement;
+            const childrenbar = bar.nextElementSibling as HTMLElement;
+
+            const clientwidth = bar.clientWidth;
+
+            const maxinpixels = translateToPixels(max,clientwidth,max,min as number);
+                    
+
+
+            finishThumb.style.left = translateToPixels(value,clientwidth,max,min as number);
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:81]: No

File: src/Inputs/Slider.tsx [hunk 12]
```diff
@@ -1,80 +1,351 @@
...
+            selectedBar.style.width = twoinputs ? (parseInt(finishThumb.style.left) - parseInt(startThumb.style.left)) +"px" :(parseInt(finishThumb.style.left)) + "px";
+
+            bar.style.width = maxinpixels;
+            childrenbar.style.width = bar.style.width;
+
+
+            const theChildren = childrenbar.children as any ;
+
+            theChildren.forEach((child:HTMLElement)=>{
+                console.log(typeof(child));
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:93]: Please remove this

File: src/Inputs/Slider.tsx [hunk 13]
```diff
@@ -1,80 +1,351 @@
...
+
+    const handleMouseDown=(e:MouseEvent)=>{
+        if (!disabled) {
+        
+          let thumb = e.target as HTMLElement;
+          let bar = thumb.parentElement as HTMLElement;
+          let selectedbar = bar.firstChild as HTMLElement;
+          let shiftX = e.clientX - thumb.getBoundingClientRect().left;
+
+          document.addEventListener('mousemove', onMouseMove as any);
```
Comment (fpunny) [inline src/Inputs/Slider.tsx:110]: Event listeners should be added and removed on component mount/unmount, not on an event

File: src/Containers/Popup.tsx (hunk 4 above)
Comment (fpunny) [inline src/Containers/Popup.tsx:13]: Also use a portal

**Comment by fpunny** [comment]
https://github.com/cheapreats/react-ui-library/issues/83

**Pull request closed without merging**